File: lib/features/logs/domain/entities/daily_log.dart
import 'package:cal_ai/features/food_recognition/domain/entities/food_analysis.dart';

class DailyLogEntity {
  final String date; // YYYY-MM-DD
  final int caloriesConsumed;
  final int carbsGrams;
  final int proteinGrams;
  final int fatsGrams;
  final List<DailyLogItemEntity> items;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  const DailyLogEntity({
    required this.date,
    required this.caloriesConsumed,
    required this.carbsGrams,
    required this.proteinGrams,
    required this.fatsGrams,
    required this.items,
    this.createdAt,
    this.updatedAt,
  });

  factory DailyLogEntity.fromJson(Map<String, dynamic> json) {
    final items = (json['items'] as List<dynamic>? ?? const [])
        .map((e) => DailyLogItemEntity.fromJson(
            e is Map<String, dynamic> ? e : <String, dynamic>{}))
        .toList();
    DateTime? parseDt(dynamic v) {
      if (v == null) return null;
      return DateTime.tryParse(v.toString());
    }

    return DailyLogEntity(
      date: (json['date'] as String).substring(0, 10),
      caloriesConsumed: (json['caloriesConsumed'] as num? ?? 0).toInt(),
      carbsGrams: (json['carbsGrams'] as num? ?? 0).toInt(),
      proteinGrams: (json['proteinGrams'] as num? ?? 0).toInt(),
      fatsGrams: (json['fatsGrams'] as num? ?? 0).toInt(),
      items: items,
      createdAt: parseDt(json['createdAt']),
      updatedAt: parseDt(json['updatedAt']),
    );
  }
}

class DailyLogItemEntity {
  final String title;
  final int calories;
  final int carbsGrams;
  final int proteinGrams;
  final int fatsGrams;
  final int? healthScore; // optional, 0..10
  final String timeIso; // ISO timestamp
  final String? imageUrl;
  final List<IngredientEntity> ingredients;

  const DailyLogItemEntity({
    required this.title,
    required this.calories,
    required this.carbsGrams,
    required this.proteinGrams,
    required this.fatsGrams,
    this.healthScore,
    required this.timeIso,
    this.imageUrl,
    this.ingredients = const [],
  });

  factory DailyLogItemEntity.fromJson(Map<String, dynamic> json) {
    int toInt(dynamic v) =>
        v is int ? v : (v is num ? v.toInt() : int.tryParse('${v ?? 0}') ?? 0);

    final ingredientsJson = json['ingredients'] as List<dynamic>? ?? [];
    final ingredients = ingredientsJson
        .map((e) => IngredientEntity.fromJson(
            e is Map<String, dynamic> ? e : <String, dynamic>{}))
        .toList();

    return DailyLogItemEntity(
      title: (json['title'] ?? '').toString(),
      calories: toInt(json['calories']),
      carbsGrams: toInt(json['carbsGrams']),
      proteinGrams: toInt(json['proteinGrams']),
      fatsGrams: toInt(json['fatsGrams']),
      healthScore: json['healthScore'] == null
          ? null
          : toInt(json['healthScore']).clamp(0, 10),
      timeIso: (json['timeIso'] ?? '').toString(),
      imageUrl: (json['imageUrl'] as String?),
      ingredients: ingredients,
    );
  }
}

File: server/src/models/DailyLog.ts
import mongoose, { Document, Schema } from 'mongoose';

export interface IIngredient {
    name: string;
    calories: number;
    proteinGrams: number;
    fatGrams: number;
    carbsGrams: number;
}

export interface ILogItem {
    title: string;
    calories: number;
    carbsGrams: number;
    proteinGrams: number;
    fatsGrams: number;
    portions?: number;
    healthScore?: number;
    timeIso: string; // ISO timestamp of when item was added
    imageUrl?: string; // optional image URL or data URI
    description?: string; // optional user-provided description for better analysis
    ingredients?: IIngredient[]; // detailed ingredients breakdown
    liked?: boolean; // whether user liked this item
}

export interface IDailyLog extends Document {
    userId: Schema.Types.ObjectId;
    date: string; // YYYY-MM-DD (local date)
    caloriesConsumed: number;
    carbsGrams: number;
    proteinGrams: number;
    fatsGrams: number;
    burnedCalories: number; // calories burned through exercise/activity
    items: ILogItem[];
    createdAt: Date;
    updatedAt: Date;
}

const logItemSchema = new Schema<ILogItem>({
    title: { type: String, required: true },
    calories: { type: Number, required: true },
    carbsGrams: { type: Number, required: true },
    proteinGrams: { type: Number, required: true },
    fatsGrams: { type: Number, required: true },
    portions: { type: Number, required: false, default: 1 },
    healthScore: { type: Number, required: false, min: 0, max: 10 },
    timeIso: { type: String, required: true },
    imageUrl: { type: String },
    description: { type: String },
    ingredients: [{
        name: { type: String, required: true },
        calories: { type: Number, required: true },
        proteinGrams: { type: Number, required: true },
        fatGrams: { type: Number, required: true },
        carbsGrams: { type: Number, required: true },
    }],
    liked: { type: Boolean, default: false },
});

const dailyLogSchema = new Schema<IDailyLog>(
    {
        userId: { type: Schema.Types.ObjectId, ref: 'User', index: true, required: true },
        date: { type: String, required: true },
        caloriesConsumed: { type: Number, required: true, default: 0 },
        carbsGrams: { type: Number, required: true, default: 0 },
        proteinGrams: { type: Number, required: true, default: 0 },
        fatsGrams: { type: Number, required: true, default: 0 },
        burnedCalories: { type: Number, required: true, default: 0 },
        items: { type: [logItemSchema], default: [] },
    },
    { timestamps: true }
);

dailyLogSchema.index({ userId: 1, date: 1 }, { unique: true });

const DailyLog = mongoose.model<IDailyLog>('DailyLog', dailyLogSchema);

export default DailyLog;


